csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.cs

```diff
@@ -101,6 +101,34 @@ public void DisableTelemetryEvents()
             NativeApiStatus.VerifySuccess(NativeMethods.OrtDisableTelemetryEvents(Handle));
         }
 
+        /// <summary>
+        /// Queries all the execution providers supported in the native onnxruntime shared library
+        /// </summary>
+        public string[] GetAvailableProviders()
+        {
+            IntPtr availableProvidersHandle = IntPtr.Zero;
+            int numProviders;
+
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtGetAvailableProviders(out availableProvidersHandle, out numProviders));
+
+            var availableProviders = new string[numProviders];
+
+            try
+            {
+                for(int i=0; i<numProviders; ++i)
+                {
+                    availableProviders[i] = NativeOnnxValueHelper.StringFromNativeUtf8(Marshal.ReadIntPtr(availableProvidersHandle, IntPtr.Size * i));
+                }
+            }
+
+            finally
+            {
+                // TODO: What if OrtReleaseAvailableProviders() throws ?
```

Comment:
>// TODO: What if OrtReleaseAvailableProviders() throws ? [](start = 16, length = 56)

First of all, C APIs never throw, it is the Verify() function that translates it into an exception. So you have an option to take action w/o exception.
I personally think, that function should not return any error, just like destructors can not return errors.
If it does, you may log it as a BUG and proceed.

Reply:
It is not a new API, hence we need to go with what we have unless we want to version the native API (which is not the scope of this PR)

Reply:
> General comment. Please, document each argument for each of the new API entry points.

Done.

Reply:
Yeah - I forgot that C APIs won't throw - they return a status.

Reply:
Added a comment based on offline discussion